atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4DecompressingInputStream.java

```diff
@@ -0,0 +1,144 @@
+    private final LZ4SafeDecompressor decompressor;
+    private final byte[] buffer;
+    private final byte[] compressedBuffer;
+    private final StreamingXXHash32 hasher;
+    private final LZ4FrameDescriptor frameDescriptor;
+    private int position;
+    private int maxPosition;
+    private int nextLength;
+
+    public LZ4DecompressingInputStream(InputStream delegate) throws IOException {
+        this.delegate = delegate;
+        this.decompressor = LZ4Streams.decompressor;
+        this.hasher = XXHashFactory.fastestInstance().newStreamingHash32(0);
+        byte[] actualMagic =
+                new byte[LZ4Streams.HEADER_SIZE + LZ4Streams.MAGIC_LENGTH + LZ4Streams.FRAME_DESCRIPTOR_SIZE];
+        this.nextLength = readBlock(actualMagic, actualMagic.length);
+        Preconditions.checkState(LZ4Streams.littleEndianIntFromBytes(actualMagic, 0) == LZ4Streams.MAGIC_VALUE,
+                "Input is not an lz4 stream that this input stream can decompress");
+        this.frameDescriptor = LZ4FrameDescriptor.fromByteArray(Arrays.copyOfRange(actualMagic,
+                LZ4Streams.MAGIC_LENGTH, LZ4Streams.MAGIC_LENGTH + LZ4Streams.FRAME_DESCRIPTOR_SIZE));
+        this.buffer = LZ4Streams.newUncompressedBuffer(frameDescriptor);
+        this.compressedBuffer = LZ4Streams.newCompressedBuffer(frameDescriptor);
+        this.position = 0;
+        this.maxPosition = 0;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if ((position == maxPosition) && !refill()) {
+            return -1;
+        }
+        return buffer[position++] & 0xff;
+    }
+
+    @Override
+    public int read(byte b[], int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException();
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+            throw new IndexOutOfBoundsException();
+        } else if (len == 0) {
+            return 0;
+        }
+
+        int bytesRead = 0;
+        while (bytesRead < len) {
+            int remainingBuffer = maxPosition - position;
+            int bytesToRead = Math.min(len - bytesRead, remainingBuffer);
+            System.arraycopy(buffer, position, b, off + bytesRead, bytesToRead);
+            position += bytesToRead;
+            bytesRead += bytesToRead;
+            if ((position == maxPosition) && !refill()) {
+                break;
+            }
+        }
+        return bytesRead > 0 ? bytesRead : -1;
+    }
+
+    @Override
+    public int available() throws IOException {
+        return maxPosition - position;
+    }
+
+    @Override
+    public void close() throws IOException {
+        delegate.close();
+    }
+
+    // This reads one block plus the size of the next block in as few
+    // read calls as possible. The length should be the number of bytes
+    // of data plus HEADER_SIZE bytes for the length of the next block.
+    private int readBlock(byte[] b, int length) throws IOException {
+        int read = 0;
```

Comment:
I'd prefer `readSoFar`, as it's difficult to infer tense from a single-word variable.